woocommerce/woocommerce-gutenberg-products-block
# Cart and Checkout block transforms for classic shortcodes

## What

This PR improves the incompatibility flow for cart and checkout blocks when non compatible extensions are detected. 

**Task List**

Adds a task list item to review the cart/checkout blocks. This notice is only visible if using blocks on the cart/checkout pages, and will redirect to either the checkout page, or cart page, on click.

![Screenshot 2023-10-17 at 14 51 37](https://github.com/woocommerce/woocommerce-blocks/assets/90977/647b2bfc-52d5-4813-82e6-39ee11389471)

**Incompatibility notices**

The notice wording is updated, and a convenient switch button is introduced to revert to classic shortcodes:

![Screenshot 2023-10-17 at 14 55 03](https://github.com/woocommerce/woocommerce-blocks/assets/90977/ffc9d426-be11-4529-b0a0-a2542d103189)

Clicking the switch button opens a confirmation model to make sure the merchant understands the ramifications of reverting to shortcodes:

![Screenshot 2023-10-17 at 14 56 33](https://github.com/woocommerce/woocommerce-blocks/assets/90977/b3afd06e-8552-438c-abf3-28d36793d274)

Confirming will replace the blocks with the classic shortcode. There is a snackbar shown giving the option to undo the action:

![Screenshot 2023-10-17 at 14 57 15](https://github.com/woocommerce/woocommerce-blocks/assets/90977/c471ee41-b7ba-4bcd-ac8f-79ff5711e1df)

**Block Transforms**

This PR also includes transforms to convert from block->shortcode even if the notice is hidden:

![Screenshot 2023-10-17 at 15 15 45](https://github.com/woocommerce/woocommerce-blocks/assets/90977/5acbb62b-e7ce-4b48-bd07-30d3ba32dba3)

Fixes https://github.com/woocommerce/woocommerce-blocks/issues/10848

## Why

This allows easier switching back to shortcodes if the block based checkout is not working or suitable for a store, and encourages merchants to check their pages are correctly configured.

## Follow ups

cc @ralucaStan Creating some follow ups so we can at least ship something in time in case these turn out to be complex tasks.

- [x] https://github.com/woocommerce/woocommerce-blocks/issues/11314
- [x] Find a way to programmatically open the block inspector based on a query string variable when navigating from the task list https://github.com/woocommerce/woocommerce-blocks/pull/11228/commits/587154835a778da20dfa35261703b4785d3b3baf

## Testing Instructions



_Please consider any edge cases this change may have, and also other areas of the product this may impact._

Prerequisite: Ensure you're already using cart and checkout blocks and you have an incompatible extension installed. You can use `WooCommerce Checkout Field Editor` as an example.

1. Go to WooCommece > Home. There should be a task list item "Review your shopper's checkout experience". Click it. You should be taken to the page editor.
2. Edit the cart page
    1. Edit the cart page
    2. Focus on the parent cart block and click its icon. You should be able to transform to "classic cart"
    3. Refresh the editor
    4. Focus on the cart block. See the incompatibility notice in the sidebar. Click the "switch to classic cart" button.
    5. Modal should open. Click switch. Block should be replaced.
    6. In the snackbar bottom left, click "undo". Should revert back to blocks.
3. Repeat above tests for **checkout block**. Pay attention to wording and ensure any references to "cart" are updated with "checkout" during the process.


## Screenshots or screencast

https://github.com/woocommerce/woocommerce-blocks/assets/90977/c923f771-cf6e-4905-8440-a70a9d2679bc

## WooCommerce Visibility


Required:


## Checklist

Required:

Conditional:

## Changelog

> Added tools to switch between Cart and Checkout Blocks and Classic Shortcodes when non-compatible payment extensions are detected.
